driver-core/src/main/java/com/datastax/driver/core/Cluster.java

```diff
@@ -1593,15 +1593,15 @@ private void connect() {
                 if (serverResponseVersion == attemptedVersion) {
                     // We're post CASSANDRA-11464: the server replies with the same version that the client tried (to
                     // ensure that the client can decode the response).
-                    // Retry with a lower version (if it's too low we'll eventually reach our min and fail).
+                    // Retry with a lower version (if it's too low we'll fail at the next attempt, as we only retry once).
```

Comment:
> the fact that retrying at N-1 is necessarily going to fail is not a documented feature

If we're going that route, the same could be said for N-2.

This is getting complicated. How about going back to the simple algorithm used [before JAVA-361](https://github.com/datastax/java-driver/blob/2.1.0/driver-core/src/main/java/com/datastax/driver/core/Cluster.java#L1143-L1168):
* try with version N
* if we get a PROTOCOL_ERROR, simply retry with the driver's previous version. Ignore the version it's encoded in (that's not documented in the spec either after all).
* repeat until we run out of driver's supported versions.

Yes, we might do a couple of extra retries in cases where there is no chance of success. It's not a big deal, negotiation is not supposed to be efficient anyway, and we discourage its use in production.